### PR TITLE
MLDB-1947 add shape and reshape builtins

### PR DESCRIPTION
**Pull request by steadtler**
Useful when dealing with tensors & tensorflow.

### CODE REVIEW

File: sql/builtin_functions.cc [hunk 1]
```diff
@@ -2629,6 +2629,83 @@ BoundFunction flatten(const std::vector<BoundSqlExpression> & args)
 
 static RegisterBuiltin registerFlatten(flatten, "flatten");
 
+BoundFunction reshape(const std::vector<BoundSqlExpression> & args)
+{
+    checkArgsSize(args.size(), 2);
+
+    if (!args[0].info->couldEmbedding())
```
Comment (jeremybarnes) [inline sql/builtin_functions.cc]: couldBeEmbedding() not couldEmbedding()

File: sql/builtin_functions.cc [hunk 2]
```diff
@@ -2629,6 +2629,83 @@ BoundFunction flatten(const std::vector<BoundSqlExpression> & args)
...
+        throw HttpReturnException(400, "requires an array as first argument");
+
+    if (!args[1].info->isEmbedding())
+        throw HttpReturnException(400, "requires an arrays as second argument");
```
Comment (jeremybarnes) [inline sql/builtin_functions.cc]: We should align with the documentation, do we call them arrays there?

File: sql/expression_value.cc [hunk 3]
```diff
@@ -561,6 +561,18 @@ EmbeddingValueInfo(std::vector<ssize_t> shape, StorageType storageType)
 {
 }
 
+EmbeddingValueInfo
```
Comment (jeremybarnes) [inline sql/expression_value.cc]: Surely this should return a shared pointer?  We basically never use bare objects for EmbeddingValueInfo.

File: sql/expression_value.h [hunk 4]
```diff
@@ -210,6 +210,12 @@ struct ExpressionValueInfo {
         return true;
     }
 
+    /// Could the thing described by this value description return an embedding?
+    virtual bool couldEmbedding() const
```
Comment (jeremybarnes) [inline sql/expression_value.h]: couldBeEmbedding()

File: sql/expression_value.h [hunk 5]
```diff
@@ -1418,6 +1434,8 @@ struct EmbeddingValueInfo: public ExpressionValueInfoT<ML::distribution<CellValu
         input. */
     EmbeddingValueInfo(const std::vector<std::shared_ptr<ExpressionValueInfo> > & input);
 
+    static EmbeddingValueInfo fromShape(const DimsVector& shape, StorageType storageType = ST_ATOM);
```
Comment (jeremybarnes) [inline sql/expression_value.h]: Again, should return a shared_ptr... and why doesn't the constructor suffice?

File: testing/MLDB-1947-reshape-builtin.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+mldb = mldb_wrapper.wrap(mldb)  # noqa
+
+class MLDB1947reshapebuiltin(MldbUnitTest):  # noqa
+
+    @classmethod
+    def setUpClass(cls):
+        pass
+
+    def test_reshape(self):
```
Comment (jeremybarnes) [inline testing/MLDB-1947-reshape-builtin.py]: Please add tests for doing it on non-embeddings.

**Comment by jeremybarnes** [comment]
Broadly OK, but a bunch of minor things to address.

File: sql/expression_value.h (hunk 5 above)
Comment (steadtler) [inline sql/expression_value.h]: In a lot of places, we call the constructor with a brace initialisation for the shape array, and its ambiguous.

**Pull request merged**